### PR TITLE
Add losses for VAE/semi-supervised CVAE

### PR DESCRIPTION
**Pull request by hsjang001205**
I research drug protein binding affinity using semi-supervised learning. 
Rewriting the model using this module, I added the losses for semi-supervised CVAE.(KL-Divergence with normal distribution, KL-Divergence, Shannon Entropy for predicted distribution)
If possible, I want to push tutorials about semi supervised learning using CVAE for this project.

### CODE REVIEW

File: deepchem/models/losses.py [hunk 1]
```diff
@@ -205,6 +205,114 @@ def loss(output, labels):
     return loss
 
 
+class VAE_ELBO(Loss):
+  """The Variational AutoEncoder loss, KL Divergence Regularize + marginal log-likelihood.
+  
+  The logvar and mu should have shape (batch_size, hidden_space).
```
Comment (rbharath) [inline deepchem/models/losses.py:211]: Could you add a references section to this docstring with links to papers introducing the ELBO loss for VAEs? Check out the numpy docs on how to add reference sections to docstring: https://numpydoc.readthedocs.io/en/latest/format.html#id8

Comment (rbharath) [reply deepchem/models/losses.py:211]: Could you also add an Examples section showing how to invoke the loss function?

File: deepchem/models/losses.py [hunk 2]
```diff
@@ -205,6 +205,114 @@ def loss(output, labels):
...
+      return BCE + kl_scale*KLD
+
+    return loss
+
+
+class VAE_KLDivergence(Loss):
+  """The KL_divergence between hidden distribution and normal distribution
+  The logvar should have shape (batch_size, hidden_space) and each term represents
+  standard deviation of hidden distribution. The mean shuold have 
+  (batch_size, hidden_space) and each term represents mean of hidden distribtuon.
```
Comment (rbharath) [inline deepchem/models/losses.py:241]: Could you add references/examples to this docstring as indicated in the comments above?

File: deepchem/models/losses.py [hunk 3]
```diff
@@ -205,6 +205,114 @@ def loss(output, labels):
...
+    def loss(logvar, mu):
+      logvar, mu = _make_pytorch_shapes_consistent(logvar, mu)
+      return 0.5 * torch.mean(torch.square(mu) + torch.square(logvar) - torch.log(1e-20 + torch.square(logvar)) - 1,-1)
+
+    return loss
+
+
+class KLDivergence(Loss):
+  """The KL_divergence between two distribution D_KL(P||Q).
+  The argument should have shape (batch_size, num of variable) and represents
```
Comment (rbharath) [inline deepchem/models/losses.py:262]: Could you add references/examples to this docstring as well?

File: deepchem/models/losses.py [hunk 4]
```diff
@@ -205,6 +205,114 @@ def loss(output, labels):
...
+        Q = torch.cat((Q,1-Q), dim = -1)
+      return torch.mean(P * torch.log((P+1e-20) / (Q+1e-20)),dim = -1)
+
+    return loss
+
+
+class ShannonEntropy(Loss):
+  """The ShannonEntropy of discrete-distribution.
+  
+  The inputs should have shape (batch size, num of variable) and represents
```
Comment (rbharath) [inline deepchem/models/losses.py:293]: Same for this docstring

**Review by rbharath — COMMENTED** [review]
Thank you for the contribution! I think this could be useful to a lot of folks. I've made a few requests for more tests and documentation that will help make this code really robust

**Comment by peastman** [comment]
TensorFlow and PyTorch both provide routines to calculate KL divergence:

https://www.tensorflow.org/api_docs/python/tf/keras/losses/KLDivergence
https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html

It's usually better to use the built in implementations directly rather than reimplementing the calculation.

**Comment by hsjang001205** [comment]
@peastman thanks for comments! I'll check it.

File: deepchem/models/losses.py [hunk 5]
```diff
@@ -205,6 +205,114 @@ def loss(output, labels):
     return loss
 
 
+class VAE_ELBO(Loss):
```
[the remark this answers is not in the document]
Comment (hsjang001205) [reply deepchem/models/losses.py:208]: All checked.

File: deepchem/models/losses.py (hunk 1 above)
Comment (hsjang001205) [reply deepchem/models/losses.py:216]: Added completed. But, I'm not sure this is a good example.

File: deepchem/models/losses.py (hunk 4 above)
Comment (hsjang001205) [reply deepchem/models/losses.py:312]: All completed.
KL_Divergence was removed because there was no difference with BCE in weight learning.

**Comment by hsjang001205** [comment]
Update/Add test_losses and reference&example.

**Review by rbharath — COMMENTED** [review]
Thanks for addressing comments! The major issue left is to fix formatting. Can you run the following commands:

```
pip install yaf==0.22
yapf --version # Check that this returns 0.22
yapf -i deepchem/models/losses.py
yapf -i deepchem/models/tests/test_losses.py
```

It looks like `yapf` is the only failing thing on the Travis tests, so once that's fixed, we should be able to merge in :)

**Comment by hsjang001205** [comment]
@rbharath Thanks :) Errors caused by my code have been resolved.

File: deepchem/models/losses.py [hunk 6]
```diff
@@ -205,6 +205,162 @@ def loss(output, labels):
...
+  """The Variational AutoEncoder loss, KL Divergence Regularize + marginal log-likelihood.
+  
+  This losses basesd on "Auto-Encoding Variational Bayes" (https://arxiv.org/abs/1312.6114).
```
Comment (rbharath) [inline deepchem/models/losses.py:211]: Could you list this in a Reference section in the docstring? Something like
```
This loss is based on [1]_.

References
-----------
.. [1] Kingma, Diederik P., and Max Welling. "Auto-encoding variational bayes." arXiv preprint arXiv:1312.6114 (2013).
```

File: deepchem/models/losses.py [hunk 7]
```diff
@@ -205,6 +205,162 @@ def loss(output, labels):
...
+      return BCE + kl_scale * KLD
+
+    return loss
+
+
+class VAE_KLDivergence(Loss):
+  """The KL_divergence between hidden distribution and normal distribution.
+  
+  This loss represents KL divergence losses between normal distribution(using parameter of distribution)
+  based on "Auto-Encoding Variational Bayes" (https://arxiv.org/abs/1312.6114).
```
Comment (rbharath) [inline deepchem/models/losses.py:269]: Could you list this in a References section as mentioned above?

File: deepchem/models/losses.py [hunk 8]
```diff
@@ -205,6 +205,162 @@ def loss(output, labels):
...
+          torch.log(1e-20 + torch.square(logvar)) - 1, -1)
+
+    return loss
+
+
+class ShannonEntropy(Loss):
+  """The ShannonEntropy of discrete-distribution.
+  
+  This loss represents shannon entropy based on
+  "A Brief Introduction to Shannon's Information Theory" (https://arxiv.org/abs/1612.09316).
```
Comment (rbharath) [inline deepchem/models/losses.py:320]: Could you list this in a References section as above?

File: deepchem/models/losses.py [hunk 9]
```diff
@@ -205,6 +205,162 @@ def loss(output, labels):
     return loss
 
 
+class VAE_ELBO(Loss):
```
Comment (rbharath) [inline deepchem/models/losses.py:208]: Could you add all these classes to the autodoc on `docs/models.rst` in the Losses section? You just need to add `autoclass` statements as for the other classes. This should hopefully be clear how to do from looking at the link here:

https://github.com/deepchem/deepchem/blob/master/docs/models.rst#L169

**Review by rbharath — COMMENTED** [review]
One last set of comments, with requests for some tweaks to the documentation. We're almost at the finish line! Once these documentation tweaks are fixed, we should be ready to merge. Thank you for your work on this :).

File: deepchem/models/losses.py (hunk 9 above)
Comment (hsjang001205) [reply deepchem/models/losses.py:208]: Finished.

**Review by rbharath — APPROVED** [review]
Looks good! Congrats on your first accepted DeepChem PR :)
